standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DirectSqlInsertPart.java

```diff
@@ -0,0 +1,804 @@
+        Object[] params = new Object[batchParamCount];
+        int index = 0;
+        int paramIndex = 0;
+        do {
+          while (index < batchRowCount && it.hasNext()) {
+            MFieldSchema fieldSchema = it.next();
+            params[paramIndex++] = cdEntry.getKey();
+            params[paramIndex++] = fieldSchema.getComment();
+            params[paramIndex++] = fieldSchema.getName();
+            params[paramIndex++] = fieldSchema.getType();
+            params[paramIndex++] = colIndex++;
+            index++;
+          }
+          if (index < batchRowCount) {
+            colIndex = 0;
+            cdEntry = cdIt.next();
+            it = cdEntry.getValue().getCols().iterator();
+          }
+        } while (index < batchRowCount);
+        executeQuery(batchQueryText, params);
+      }
+    };
+    insertInBatch("\"COLUMNS_V2\"", columns, columnCount, rowCount, batchExecutionContext);
+  }
+
+  private void insertBucketColInBatch(Map<Long, MStorageDescriptor> sdIdToStorageDescriptor) throws MetaException {
+    int rowCount = 0;
+    for (MStorageDescriptor sd : sdIdToStorageDescriptor.values()) {
+      rowCount += sd.getBucketCols().size();
+    }
+    if (rowCount == 0) {
+      return;
+    }
+    String columns = "(\"SD_ID\",\"BUCKET_COL_NAME\",\"INTEGER_IDX\")";
+    int columnCount = 3;
+    BatchExecutionContext batchExecutionContext = new BatchExecutionContext() {
+      int colIndex = 0;
+      final Iterator<Map.Entry<Long, MStorageDescriptor>> sdIt = sdIdToStorageDescriptor.entrySet().iterator();
+      Map.Entry<Long, MStorageDescriptor> sdEntry = sdIt.next();
+      Iterator<String> it = sdEntry.getValue().getBucketCols().iterator();
+      @Override
+      public void execute(String batchQueryText, int batchRowCount, int batchParamCount) throws MetaException {
+        Object[] params = new Object[batchParamCount];
+        int index = 0;
+        int paramIndex = 0;
+        do {
+          while (index < batchRowCount && it.hasNext()) {
+            params[paramIndex++] = sdEntry.getKey();
+            params[paramIndex++] = it.next();
+            params[paramIndex++] = colIndex++;
+            index++;
+          }
+          if (index < batchRowCount) {
+            colIndex = 0;
+            sdEntry = sdIt.next();
+            it = sdEntry.getValue().getBucketCols().iterator();
+          }
+        } while (index < batchRowCount);
+        executeQuery(batchQueryText, params);
+      }
+    };
+    insertInBatch("\"BUCKETING_COLS\"", columns, columnCount, rowCount, batchExecutionContext);
+  }
+
+  private void insertSortColInBatch(Map<Long, MStorageDescriptor> sdIdToStorageDescriptor) throws MetaException {
+    int rowCount = 0;
+    for (MStorageDescriptor sd : sdIdToStorageDescriptor.values()) {
+      rowCount += sd.getSortCols().size();
+    }
+    if (rowCount == 0) {
+      return;
+    }
+    String columns = "(\"SD_ID\",\"COLUMN_NAME\",\"ORDER\",\"INTEGER_IDX\")";
+    int columnCount = 4;
```

Comment:
How about iterate the order by this way:
```java
Iterator<Object[]> iterator = new Iterator<Object[]>() {
      final Iterator<Map.Entry<Long, MStorageDescriptor>> sdIt = sdIdToStorageDescriptor.entrySet().iterator();
      Iterator<MOrder> iter = null;
      Long key = null;
      int idx = 0;
      @Override
      public boolean hasNext() {
        while ((iter == null || !iter.hasNext()) && sdIt.hasNext()) {
          Map.Entry<Long, MStorageDescriptor> entry = sdIt.next();
          if (entry.getValue().getSortCols() != null) {
            iter = entry.getValue().getSortCols().iterator();
            key = entry.getKey();
            idx = 0;
          }
        }
        return iter != null && iter.hasNext();
      }

      @Override
      public Object[] next() {
        Object[] params = new Object[4];
        MOrder order = iter.next();
        params[0] = key;
        params[1] = order.getCol();
        params[2] = order.getOrder();
        params[3] = idx++;
        return params;
      }
    };

Reply:
Actually have optimized to reduce the redundant checks and filled them with single params array. Any specific reason to suggest the change?